teps/0118-matrix-with-explicit-combinations-of-parameters.md

```diff
@@ -0,0 +1,415 @@
+    - ...
+```
+
+`Matrix` by design with multiple `Parameters` calculates all the possible combinations and executes an instance of
+`PipelineTask` with each combination. Now, this specification of the `Matrix` creates 9 different combinations, such
+as:
+
+![Matrix for Kaniko with multiple Images and Dockerfile](images/0118-matrix-kaniko-with-images-and-dockerfile.jpg)
+
+This is not a desired outcome either. We need a way to specify explicit mapping between `IMAGE` and `DOCKERFILE`, such
+as:
+
+```yaml
+    - IMAGE: "image-1"
+      DOCKERFILE: "path/to/Dockerfile1"
+
+    - IMAGE: "image-2"
+      DOCKERFILE: "path/to/Dockerfile2"
+
+    - IMAGE: "image-3"
+      DOCKERFILE: "path/to/Dockerfile3"
+```
+
+### Goals
+
+The main goal of this TEP is to extend `Matrix` functionality to enable executing a `PipelineTask` with a limited or
+specific set of combinations of `Parameters`.
+
+### Non-Goals
+
+This TEP focuses on adding a specific combination of input values for `Matrix` parameters, excluding a specific
+combination from `Matrix` is out of scope for this TEP.
+
+## Proposal
+
+We propose adding a field - `include` - within the `matrix` field. This requires that we also move the existing array
+of `Parameters` to a field - `params` - within the `matrix` field. This allows users to specify the `Parameters` 
+whose combinations should be generated, and the explicit combinations of `Parameters` that need to be added into the
+`Matrix`.
+
+#### Include specific combinations in the Matrix
+
+Take the use case described [above](#adding-a-specific-combination-in-the-matrix). The `Pipeline` will need to define a 
+`Matrix` that will execute `golang-test` in three different linux flavors. While running on `amd64` and `ppc64le`,
+`golang-test` should execute `go test github.com/tektoncd/pipeline/pkg/... -race -cover -v`. While running on 
+`linux/s390x`, `golang-test` should execute `go test github.com/tektoncd/pipeline/pkg/... -cover -v`.
+
+The `Pipeline` addressing this use case will be defined as shown below:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline-to-build-and-test-go-project
+spec:
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: golang-test
+      taskRef:
+        name: golang-test
+      workspaces:
+        - name: source
+          workspace: shared-workspace
+      params:
+        - name: package
+          value: "github.com/tektoncd/pipeline"
+        - name: packages
+          value: "./pkg/..."
+      matrix:
+        params:
+          - name: GOARCH
+            value:
+              - "linux/amd64"
+              - "linux/ppc64le"
+          - name: flags
+            value:
+              - "-race -cover -v"
+        include:
+          - name: s390x-no-race 
+            params:
+              - name: GOARCH
+                value: "linux/s390x"
+              - name: flags
+                value: "-cover -v" 
+    - ...
```

Comment:
Let's say you wanted to matrix an additional parameter here, e.g. you want the following combos:
- {"linux/amd64", "-race -cover -v", "chrome"}
- {"linux/ppc64le", "-race -cover -v", "chrome"}
- {"linux/s390x", "-cover -v", "chrome }
- {"linux/amd64", "-race -cover -v", "safari"}
- {"linux/ppc64le", "-race -cover -v", "safari"}
- {"linux/s390x", "-cover -v", "safari"}

Would you write your pipeline like this?
```
      matrix:
        params:
          - name: GOARCH
            value:
              - "linux/amd64"
              - "linux/ppc64le"
          - name: flags
            value:
              - "-race -cover -v"
          - name: browser
            value:
            - "chrome"
            - "safari"
        include:
          - name: s390x-no-race-chrome
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags
                value: "-cover -v" 
              - name: browser
                value: "chrome"
          - name: s390x-no-race-safari
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags
                value: "-cover -v" 
              - name: browser
                value: safari
```

or like this?
```
      matrix:
        params:
          - name: GOARCH
            value:
              - "linux/amd64"
              - "linux/ppc64le"
          - name: flags
            value:
              - "-race -cover -v"
          - name: browser
            value:
            - "chrome"
            - "safari"
        include:
          - name: s390x-no-race
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags
                value: "-cover -v"
```

Reply:
the first one is correct

the second one will create these combinations:
- {"linux/amd64", "-race -cover -v", "chrome"}
- {"linux/ppc64le", "-race -cover -v", "chrome"}
- {"linux/amd64", "-race -cover -v", "safari"}
- {"linux/ppc64le", "-race -cover -v", "safari"}
- {"linux/s390x", "-cover -v"}

Reply:
Got it! Do you know of any reasons to prefer one or the other? I'm sure you're already aware that [GH Actions](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations) uses the same logic you're proposing, and I'm curious why they made that decision.

IMO, the syntax 
```
 - name: s390x-no-race-chrome
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags
                value: "-cover -v" 
              - name: browser
                value: "chrome"
          - name: s390x-no-race-safari
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags
                value: "-cover -v" 
              - name: browser
                value: safari
```
feels a bit like why we introduced matrix in the first place; it feels nicer to not have to write out combinations for the third param "browser".

However, if we did decide to go with the second version in my example, it's not clear how you would generate the combination `{"linux/s390x", "-cover -v"}` unless you could specify "none" as a param value, e.g.
```
  include:
          - name: s390x-no-race
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags
                value: "-cover -v"
              - name: browser
                value: None   # or similar
```
WDYT?

Reply:
Hey @lbernick to get the following combos (copied from your initial comment for readability):

- {"linux/amd64", "-race -cover -v", "chrome"}
- {"linux/ppc64le", "-race -cover -v", "chrome"}
- {"linux/s390x", "-cover -v", "chrome }
- {"linux/amd64", "-race -cover -v", "safari"}
- {"linux/ppc64le", "-race -cover -v", "safari"}
- {"linux/s390x", "-cover -v", "safari"}


This combo includes three params `GOARCH`, `flags`, `browser` (which is a little weird example while talking about golang test but its fine to include from the understanding perspective).

The matrix `params` section will include `GOARCH` and `browser` which will result in six combinations. The `include` section will expand on the each combination based on the `flags`.

You will write your pipeline like this:

```
      matrix:
        params:
          - name: GOARCH
            value:
              - "linux/amd64"
              - "linux/ppc64le"
              - "linux/s390x"
          - name: browser
            value:
            - "chrome"
            - "safari"
        include:
          - name: s390x-no-race
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags
                value: "-cover -v" 
```

Here we are leveraging default value of `flags` which is `"-race -cover -v"`. Let's say there was no default value for flags available, we will define the pipeline like this:

```
      matrix:
        params:
          - name: GOARCH
            value:
              - "linux/amd64"
              - "linux/ppc64le"
              - "linux/s390x"
          - name: browser
             value:
               - "chrome"
               - "safari"
          - name: flags
             value: "-race -cover -v" 
        include:
          - name: s390x-no-race
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags
                value: "-cover -v"
```

Reply:
I have updated the original [proposal](https://hackmd.io/@r5gvA6KSS4mn5j_dy4QR9w/B1BD7cyn9/edit) (waiting for @jerop to update this PR 🤗).

Original:

```
matrix:
        params:
          - name: GOARCH
            value:
              - "linux/amd64"
              - "linux/ppc64le"
          - name: flags
            value:
              - "-race -cover -v"
        include:
          - name: s390x-no-race 
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags
                value: "-cover -v" 
```

Updated:

```
 matrix:
        params:
          - name: GOARCH
            value:
              - "linux/amd64"
              - "linux/ppc64le"
              - "linux/s390x"
        include:
          - name: s390x-no-race 
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags
                value: "-cover -v" 
```